### ISSUE TITLE
LB-Service doesn't response after master shutdown

### ISSUE THREAD

**Issue by runam0K** [google/metallb#237]
**Is this a bug report or a feature request?**:
Bug?

**What happened**:
When I shut down my kubernetes-master, the loadbalancer-service isnt responding after about four minutes. Everything works before that time.

If I start the master again then I can access the service again.

Is this a normal behavior? How can I solve the problem?

**What you expected to happen**:
The service should continue to be accessible.

**How to reproduce it (as minimally and precisely as possible)**:
- Shut down master
- Wait for four minutes
- Service is not responsible anymore

**Anything else we need to know?**:
ConfigMap:
```
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: metallb-system
  name: config
data:
  config: |
    address-pools:
    - name: default
      protocol: layer2
      addresses:
      - 10.42.20.80-10.42.20.90
```

LB Service:
```
apiVersion: v1
kind: Service
metadata:
  name: ingress-nginx-lb
  namespace: ingress-nginx
spec:
  type: LoadBalancer
  ports:
  - name: http
    port: 80
  - name: https
    port: 443
  selector:
    app: ingress-nginx
```

**Environment**:
- MetalLB version: latest
- Kubernetes version: 1.10
- BGP router type/version:
- OS (e.g. from /etc/os-release): debian stretch
- Kernel (e.g. `uname -a`): 4.9.0-6-amd64

**Comment by danderson** [comment]
This is currently expected with layer2 mode. Layer 2 mode uses Kubernetes leader election to select a leader node that owns all the load-balancer IPs. This leadership election depends on control plane availability, because the leader has to renew its lease every 10s to remain the leader.

When the control plane goes down, leadership is lost almost immediately (<10s). The reason it still seems to work for several minutes is actually just the ARP cache on your client, which persists for some number of minutes even though there are no more ARP responses on the network.

#195 will address this by removing leader election from layer2 mode, and switching to a distributed mechanism that can survive the control plane going down. It's also a tradeoff because it can lead to split brain issues in some cases (the fundamental tradeoff of distributed computing - consistency, or availability), but I think it will be a better tradeoff for MetalLB.

**Comment by runam0K** [comment]
Thank you for the feedback and explanation.

If I understand correctly # 195 will be implemented soon?
Currently, it would be the best alternative to use BGP or install a second master for HA to work around this problem?

**Comment by danderson** [comment]
I can't guarantee any implementation timelines, because I'm basically the only one implementing large features right now, and I'm doing it in my spare time. But, #195 is reasonably high up in my wishlist right now.

Currently, the best workarounds for this are to use an HA control plane for k8s (you need 3 masters for this, because the HA control plane requires a majority of machines to be up - so with 2 machines you still can't lose 1 machine). Alternatively, BGP mode should be resistant to control plane outages, because the only dependency is for _changes_ to the LB setup, unlike layer2 where master election is a continuous requirement.

**Comment by danderson** [comment]
Closing, since this is basically asking for #195 to be implemented, and #195 is already tracking that work by itself.
